c++/include/orc/Writer.hh

```diff
@@ -0,0 +1,294 @@
+    double getDictionaryKeySizeThreshold() const;
+
+    /**
+     * Set whether or not to have block padding.
+     */
+    WriterOptions& setBlockPadding(bool padding);
+
+    /**
+     * Get whether or not to have block padding.
+     * @return if not set, return default value which is false.
+     */
+    bool getBlockPadding() const;
+
+    /**
+     * Set Run length encoding version
+     */
+    WriterOptions& setRleVersion(RleVersion version);
+
+    /**
+     * Get Run Length Encoding version
+     */
+    RleVersion getRleVersion() const;
+
+    /**
+     * Set compression kind.
+     */
+    WriterOptions& setCompression(CompressionKind comp);
+
+    /**
+     * Get the compression kind.
+     * @return if not set, return default value which is ZLIB.
+     */
+    CompressionKind getCompression() const;
+
+    /**
+     * Set the encoding strategy.
+     */
+    WriterOptions& setEncodingStrategy(EncodingStrategy strategy);
+
+    /**
+     * Get the encoding strategy.
+     * @return if not set, return default value which is SPEED.
+     */
+    EncodingStrategy getEncodingStrategy() const;
+
+    /**
+     * Set the compression strategy.
+     */
+    WriterOptions& setCompressionStrategy(CompressionStrategy strategy);
+
+    /**
+     * Get the compression strategy.
+     * @return if not set, return default value which is speed.
+     */
+    CompressionStrategy getCompressionStrategy() const;
+
+    /**
+     * Set the writer version.
+     */
+    WriterOptions& setWriterVersion(WriterVersion version);
+
+    /**
+     * Get the strip size.
+     * @return if not set, return default value which is WriterVersion_ORIGINAL
+     */
+    WriterVersion getWriterVersion() const;
+
+    /**
+     * Set the padding tolerance.
+     */
+    WriterOptions& setPaddingTolerance(double tolerance);
+
+    /**
+     * Get the padding tolerance.
+     * @return if not set, return default value which is zero.
+     */
+    double getPaddingTolerance() const;
+
+    /**
+     * Set the memory pool.
+     */
+    WriterOptions& setMemoryPool(MemoryPool * memoryPool);
+
+    /**
+     * Get the strip size.
+     * @return if not set, return default memory pool.
+     */
+    MemoryPool * getMemoryPool() const;
+
+    /**
+     * Set the error stream.
+     */
+    WriterOptions& setErrorStream(std::ostream& errStream);
+
+    /**
+     * Get the error stream.
+     * @return if not set, return std::err.
+     */
+    std::ostream * getErrorStream() const;
+
+    /**
+     * Set whether or not to write statistics (file statistics,
+     * stripe statistics, etc.)
+     */
+    WriterOptions& setEnableStats(bool enable);
+
+    /**
+     * Get whether or not to write statistics (file statistics,
+     * stripe statistics, etc.)
+     * @return if not set, the default is true
+     */
+     bool getEnableStats() const;
+
+    /**
+     * Set whether or not to update string min/max stats
+     */
+    WriterOptions& setEnableStrStatsCmp(bool enable);
```

Comment:
The memcpy isn't that expensive since it only happens when there is a new minimum or maximum. This would require a change to the ORC specification and would likely break current readers. We can discuss making such a change after we have performance numbers, but I really don't expect computing string statistics to be a noticeable cost.